### ISSUE TITLE
Question: Scaling data when using LightGBM cross-validation

### ISSUE THREAD

**Issue by Dr4cky** [microsoft/LightGBM#4500]
I was wondering whether it is possible to scale data when using the LightGBM cross-validation. So far, it seems that I have to program the cross-validation manually such that I can standardize the training data and apply the same transformation to the validation data in each iteration.

Am I missing something, or is it simply not possible?

Thanks!

**Comment by jmoralez** [comment]
Keep in mind that scaling the features shouldn't have any effect on tree-based models like lightgbm, however if you want to do it you can use the scikit-learn API with something like the following:
```python
import lightgbm as lgb
import pandas as pd
from sklearn.datasets import make_regression
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

X, y = make_regression(n_samples=1_000, n_features=4, n_informative=2, random_state=0)
estimator = Pipeline([
    ('scaler', 'passthrough'),
    ('model', lgb.LGBMRegressor(random_state=0))
])
search = GridSearchCV(
    estimator,
    {'scaler': ['passthrough', StandardScaler()], 'model__': [0.1, 0.01]},
    scoring='neg_mean_squared_error',
)
search.fit(X, y)
search_df = pd.DataFrame(search.cv_results_['params'])
search_df['score'] = search.cv_results_['mean_test_score']
search_df
```
|   model__learning_rate | scaler           |      score |
|-----------------------:|:-----------------|-----------:|
|                   0.1  | passthrough      |   -92.827  |
|                   0.1  | StandardScaler() |   -94.8684 |
|                   0.01 | passthrough      | -1311.86   |
|                   0.01 | StandardScaler() | -1312.23   |

**Comment by jmoralez** [comment]
@jameslamb do you know why the scores aren't identical? I just tried this example with xgboost and `sklearn.ensemble.GradientBoostingRegressor` and they get the exact same scores with and without scaling.

**Comment by StrikerRUS** [comment]
@Dr4cky 
> I was wondering whether it is possible to scale data when using the LightGBM cross-validation.

I guess you need `fpreproc` argument.
https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.cv.html
https://github.com/microsoft/LightGBM/blob/d05f54701ee88a1373bd9e0a1ea4317827015c9c/tests/python_package_test/test_engine.py#L1589-L1607

**Comment by lcrmorin** [comment]
@jmoralez I came here because I have a similar question. I get that for tree models the scale shouldn't have an impact (see the discussion here https://datascience.stackexchange.com/questions/60950/is-it-necessary-to-normalize-data-for-xgboost for example). However I don't really see how it would hold with l1 or l2 regularisation. Scale usually matter a lot when there is regularisation. For exemple L1 regularisation might drop features with more important scales first. I don't really understand why 'scaling data is not necessary' would hold when there is regularisation in GBT models. This might exeplain the difference here if there is a small regularisation enabled by default.

**Comment by jmoralez** [comment]
I think you may be thinking of L1 and L2 regularization in the context of penalizing coefficients as with Lasso or Ridge. In GBDT the regularization is applied to the split gain. So for example if you have a split at `x = 100` with some gain `g` and you try scaling your feature, the split may now be at `x = 50` but the gain should be the same.

**Comment by lcrmorin** [comment]
Yes I was interpreting the l1-l2 penalisation as penalising coefficients. Do you have a source that explain the penalisation of the gain in details ? I have found : [Regularized Gradient Boosting](https://papers.nips.cc/paper/2019/file/465636eb4a7ff4b267f3b765d07a02da-Paper.pdf) but it is a bit stale in terms of illustrations.

**Comment by jmoralez** [comment]
Hi @lcrmorin, sorry for the late response.

In LightGBM there are two parameters that control the regularization: `reg_alpha` and `reg_lambda` and they roughly work like in the following equation:
<a href="https://www.codecogs.com/eqnedit.php?latex=SplitGain&space;=&space;\frac{(\sum_{i&space;\in&space;I_L}g_i-\alpha)^2}{\sum_{i&space;\in&space;I_L}h_i&space;&plus;&space;\lambda}&space;&plus;&space;\frac{(\sum_{i&space;\in&space;I_R}g_i-\alpha)^2}{\sum_{i&space;\in&space;I_R}h_i&space;&plus;&space;\lambda}&space;-&space;\frac{(\sum_{i&space;\in&space;I}g_i-\alpha)^2}{\sum_{i&space;\in&space;I}h_i&space;&plus;&space;\lambda}" target="_blank"><img src="https://latex.codecogs.com/svg.latex?SplitGain&space;=&space;\frac{(\sum_{i&space;\in&space;I_L}g_i-\alpha)^2}{\sum_{i&space;\in&space;I_L}h_i&space;&plus;&space;\lambda}&space;&plus;&space;\frac{(\sum_{i&space;\in&space;I_R}g_i-\alpha)^2}{\sum_{i&space;\in&space;I_R}h_i&space;&plus;&space;\lambda}&space;-&space;\frac{(\sum_{i&space;\in&space;I}g_i-\alpha)^2}{\sum_{i&space;\in&space;I}h_i&space;&plus;&space;\lambda}" title="SplitGain = \frac{(\sum_{i \in I_L}g_i-\alpha)^2}{\sum_{i \in I_L}h_i + \lambda} + \frac{(\sum_{i \in I_R}g_i-\alpha)^2}{\sum_{i \in I_R}h_i + \lambda} - \frac{(\sum_{i \in I}g_i-\alpha)^2}{\sum_{i \in I}h_i + \lambda}" /></a>

Note that I said roughly because the numerators are actually lower bounded by zero. So as you can see `reg_alpha` can turn a small gain into zero, as in L1 regularization.

On the other hand `reg_lambda` acts on the denominator, making the gain smaller in a smoother way but hardly making it zero, as in L2 regularization.

Here's a small example showing this for the L2 objective (mean squared error), where the gradients are `y - pred` and the hessians are ones:

```python
import lightgbm as lgb
import numpy as np

def get_split_gain(x, gradients, threshold, l1=0, l2=0):
    gs = gradients.sum()
    hs = gradients.size
    left_mask = x <= threshold
    right_mask = ~left_mask
    gls = gradients[left_mask].sum()
    grs = gradients[right_mask].sum()
    hls = left_mask.sum()
    hrs = right_mask.sum()
    return (gls - l1)**2 / (hls + l2) + (grs - l1)**2 / (hrs + l2) - (gs - l1)**2 / (hs + l2)

def get_threshold_and_gain_from_model(X, y, l1=0, l2=0):
    model = lgb.LGBMRegressor(n_estimators=1, num_leaves=2, reg_alpha=l1, reg_lambda=l2)
    model.fit(X, y, init_score=np.zeros(y.size))
    trees_df = model.booster_.trees_to_dataframe()
    return trees_df.loc[trees_df.node_depth == 1, ['threshold', 'split_gain']].iloc[0]    

np.random.seed(0)
x = np.linspace(-1, 1, 100)
X = x.reshape(-1, 1)
y = x**2 + 0.1*np.random.rand(x.size)

threshold, gain = get_threshold_and_gain_from_model(X, y)
threshold, gain
# (-0.5858585858585857, 2.7980101108551025)

gradients = y # init_score is 0
get_split_gain(x, gradients, threshold)
# 2.798012322171269

# L1
_, gain = get_threshold_and_gain_from_model(X, y, l1=0.1)
# 2.673409938812256
get_split_gain(x, gradients, threshold, l1=0.1)
# 2.673407082375469

# L2
_, gain = get_threshold_and_gain_from_model(X, y, l2=0.1)
# 2.7535200119018555
get_split_gain(x, gradients, threshold, l2=0.1)
# 2.753522844719676
```
Note that this only involves the target (through the gradients) so scaling the features shouldn't have any effect on the gain, even with regularization. Hope this gives you a better idea of how the regularization works.
